Zend/zend_opcode.c

```diff
@@ -153,6 +153,12 @@ ZEND_API void zend_function_dtor(zval *zv)
 		ZEND_ASSERT(function->common.function_name);
 		zend_string_release_ex(function->common.function_name, 1);
 
+		if (function->common.attributes) {
```

Comment:
There's at least one bug in method inheritance and it's probably that for ext/pdo/pdo_dbh.c. A base class and its subclass would have attributes pointing to the same thing, and I think it should only be freed once, on the base class for internal functions.

- Userland isn't affected because it's allocated on the arena, so there's nothing to free. Only need it for `function->type == ZEND_INTERNAL_FUNCTION`

`function->common.scope` is non-null for methods and null for functions.

Instead of/alongside `fn->common.fn_flags & (ZEND_ACC_HAS_RETURN_TYPE|ZEND_ACC_HAS_TYPE_HINTS)) && fn->common.scope == ce`, this would check `fn->common.attributes && fn->common.scope == ce` in a solution, I believe

```c
		/* For methods this will be called explicitly. */
		if (!function->common.scope) {
			zend_free_internal_arg_info(&function->internal_function);
		}
```

```c
		case ZEND_INTERNAL_CLASS:
// ....
			/* TODO: eliminate this loop for classes without functions with arg_info */
			ZEND_HASH_MAP_FOREACH_PTR(&ce->function_table, fn) {
				if ((fn->common.fn_flags & (ZEND_ACC_HAS_RETURN_TYPE|ZEND_ACC_HAS_TYPE_HINTS)) &&
				    fn->common.scope == ce) {
					zend_free_internal_arg_info(&fn->internal_function);
				}
			} ZEND_HASH_FOREACH_END();
```

Reply:
@TysonAndre Thanks for looking into this. I'm afraid I don't follow, because my internals expertise is very much limited.

> There's at least one bug in method inheritance (hopefully just that one) and it's probably that for ext/pdo/pdo_dbh.c.

pdo_dbh.c is not part of this PR. It's part of my RFC PR. Did you mix those up or are you referring to an existing bug?

With this PR `.attributes` should only be non-NULL for the non-method function `zend_test_parameter_with_attribute`.

Reply:
> pdo_dbh.c is not part of this PR. It's part of my RFC PR. Did you mix those up or are you referring to an existing bug?

Oops, I mixed those up, but I think that's also a real bug. This would be a new type of double free that isn't tested by your test (would need to add a method with a function attribute to trigger that).

> With this PR `.attributes` should only be non-NULL for the non-method function `zend_test_parameter_with_attribute`.

Okay, so I'm not sure. I'd have to compile this and see where the double free was happening